crates/notion-core/src/tool/mod.rs

```diff
@@ -123,6 +125,35 @@ impl ToolSpec {
             package => ToolSpec::Package(package.into(), version),
         })
     }
+
+    /// Check the args for the bad pattern of `notion install <tool> <number>`.
+    ///
+    /// Requires you to pass *all* args. Should only be called from commands
+    /// like `notion install`, `notion fetch`, or `notion pin`.
+    pub fn check_args<I>(mut args: I, action: String) -> Fallible<()>
+    where
+        I: Iterator<Item = String>,
+    {
+        // The case we are concerned with is where we have `<tool> <number>`.
+        // Those arguments will always follow `notion` and one of `pin`,
+        // `fetch`, or `install`. We use `nth(2)` to advance the iterator to the
+        // first item of interest, then `next` to get the other item.
+        if let (Some(name), Some(maybe_version)) = (args.nth(2), args.next()) {
```

Comment:
Using specific positioning could pose a problem, as we have the `--verbose` flag that could ruin this logic, e.g. `notion --verbose install node 10` wouldn't cause an error because it's reading `install` and `node` instead of `node` and `10`.

It may be better to use the tools vector from the commands, since that has already been parsed by StructOpt.

Additionally, we need to make sure we only error if there are exactly 2 parameters. If someone does `notion install node 10 yarn`, I think we can let that through since at that point they know they're installing multiple different things.

Reply:
Ah, blast, that's a good point re: flags. Unfortunately, the `tools` vector for the commands *can't* be used here: we've lost the relevant information by that point—namely, what the bare commands were. We can attach that, perhaps, but they're parsed into the relevant data structures by that point.

Reply:
Ah, good point about not being able to use the tools, since once it’s parsed you can’t tell the difference between `node` and `node@lts`.

A couple ideas I had for working around that:

1. Change all of the StructOpt types to take a `Vec<String>`, then add a helper function (maybe in `command/mod.rs`) that accepts that vector and does this verification, the `ToolSpec` parsing, and the sorting. Then each command can call that function and use the results from there.

2. Use `.filter()` on the iterator of args, removing any args that begin with a `-`, then use the positional logic on that.

The latter is the approach we use in `notion-core/src/tool/yarn.rs` to detect global installs, since those are also positional (the first two non-flag arguments need to be `global install` in that order)

Reply:
I keep going back and forth. If we ever introduce an *optional argument* command specified like `--foo=true`, (2) breaks. At the same time, I *really* don't want to pull the parsing behavior out of the actual CLI parsing; it's *really* nice that if you have a command, it's already a valid set of `tools` and there's no further work to be done. 🤔

Reply:
Per our brief discussion on Discord, I ended up deciding that the best and cleanest thing was to wrap all of these bits into a `ToolSpec::from_strings` function which returns `Fallible<Vec<ToolSpec>>` – this worked out nicely and lets us just do this in the actual tool invocations:

```rust
for tool in ToolSpec::from_strings(&self.tools, "install")? {
  // ...
}
```

This includes being able to close over the need to make sure the tools are all sorted as their final operation, which is a nice win for maintainability.